BuildScript.UnitTests/FilterTestConfigurationsTest.cs

```diff
@@ -0,0 +1,498 @@
+    {
+      const string unsupportedDatabaseSystem = "SomeUnsupportedDatabaseSystem";
+      var testConfiguration = CreateTestConfiguration (databaseSystem: unsupportedDatabaseSystem);
+      var testConfigurations = new[] { testConfiguration };
+      var supportedDatabaseSystems = new[] { "SqlServer2014" };
+      var loggerMock = MockRepository.GenerateMock<ITaskLogger>();
+      loggerMock.Expect (
+          _ => _.LogError ($"Cannot filter with unsupported {TestingConfigurationMetadata.DatabaseSystem} '{unsupportedDatabaseSystem}'"));
+      var filter = CreateFilterTestConfigurations (supportedDatabaseSystems: supportedDatabaseSystems, logger: loggerMock);
+
+      Assert.That (() => filter.GetFilteredTestConfigurations (testConfigurations), Throws.Exception);
+      loggerMock.VerifyAllExpectations();
+    }
+
+    [Test]
+    public void SupportedExecutionRuntimes_FilterForUnsupportedExecutionRuntimes_Error ()
+    {
+      const string unsupportedExecutionRuntime = "SomeUnsupportedExecutionRuntime";
+      var testConfiguration = CreateTestConfiguration (executionRuntime: unsupportedExecutionRuntime);
+      var testConfigurations = new[] { testConfiguration };
+      var supportedExecutionRuntimes = new[] { "Win_NET48::DockerImageForWinNET48" };
+      var loggerMock = MockRepository.GenerateMock<ITaskLogger>();
+      loggerMock.Expect (
+          _ => _.LogError ($"Cannot filter with unsupported {TestingConfigurationMetadata.ExecutionRuntime} '{unsupportedExecutionRuntime}'"));
+      var filter = CreateFilterTestConfigurations (supportedExecutionRuntimes: supportedExecutionRuntimes, logger: loggerMock);
+
+      Assert.That (() => filter.GetFilteredTestConfigurations (testConfigurations), Throws.Exception);
+      loggerMock.VerifyAllExpectations();
+    }
+
+    [Test]
+    public void SupportedExecutionRuntimes_LocalMachineNotEnabled ()
+    {
+      var testConfiguration = CreateTestConfiguration (executionRuntime: "LocalMachine");
+      var testConfigurations = new[] { testConfiguration, _omnipresentTestConfiguration };
+      var supportedExecutionRuntimes = new[] { "Win_NET45::DockerImageForWinNET45" };
+      var allExecutionRuntimes = new[] { "Win_NET45" };
+      var filter = CreateFilterTestConfigurations (
+          supportedExecutionRuntimes: supportedExecutionRuntimes,
+          enabledExecutionRuntimes: allExecutionRuntimes);
+
+      var filteredTestConfigurations = filter.GetFilteredTestConfigurations (testConfigurations);
+
+      Assert.That (filteredTestConfigurations, Is.EquivalentTo (new[] { _omnipresentTestConfiguration }));
+    }
+
+    [Test]
+    public void SupportedExecutionRuntimes_SameImageWithDifferentKeys_Error ()
+    {
+      var supportedExecutionRuntimes = new[] { "Key1::SameImage", "Key2::SameImage" };
+      var loggerMock = MockRepository.GenerateMock<ITaskLogger>();
+      loggerMock.Expect (_ => _.LogError ("It is not possible to assign a docker image to multiple keys."));
```

Comment:
This is something that came up while testing. The `TestConfigurationsFilter` needs to perform reverse lookups too since the `TestConiguration` objects carry the DockerImageName as their ExecutionRuntime property but the `EnabledExecutionRuntimes` are the keys to the docker image names, meaning we need bijective mapping for validation